### ISSUE TITLE
.npy file instead of tiff images

### ISSUE THREAD

**Issue by nayemabs** [azavea/raster-vision#1392]
Can I use .npy files instead of .tiff as inputs to the segmentation model? I have multiband/multichannel fusion images that are in the .npy format.

**Comment by AdeelH** [comment]
Are these large rasters that need to be broken up into smaller chips before being fed into the model or are these small chips?

If the former, that is currently not supported but should be very simple to implement. Essentially, we'll need to implement a `NumpySource` analogous to the existing [`RasterioSource`](https://github.com/azavea/raster-vision/blob/master/rastervision_core/rastervision/core/data/raster_source/rasterio_source.py) that handles tiff files.

**Comment by nayemabs** [comment]
Its already in smaller chips format.

**Comment by AdeelH** [comment]
That is supported, but perhaps not in the most convenient way.

Can you tell me a bit more about how the files are organized? Are the chips grouped by geographical "scenes" e.g. where each scene is a different city etc. or are they just a raw collection of chips?

**Comment by nayemabs** [comment]
They are chips of the same city. Chopped off from top left to bottom right in order of their naming from the original full-sized tile. The filenames are: 2019-09-07_0000000001-2019-09-07_0000000700 where there are 700 chips of size 512x512. Basically, multiple bands of satellite data have been stacked with np.dstack and saved to .npy format.  

Alternatively, saving the multichannel data to .tiff has resulted in weird formats where QGI hangs up to open them. Thus the .npy format.

**Comment by AdeelH** [comment]
Here is a minimal example. It assumes that we're using the current master branch.

To test it, I created some random 13-band chips along with their corresponding labels like so:
```python
import numpy as np

for split in ['train', 'valid']:
    for i in range(10):
        img = np.random.random(size=(300, 300, 13))
        label = (img[..., 0] > 0.5).astype(int)
        np.save(f'data/{split}/img/{i:03d}.npy', img)
        np.save(f'data/{split}/labels/{i:03d}.npy', label)
```
This produces the following directory structure. This structure is what Raster Vision expects.
```
data
├── train
│   ├── img
│   │   ├── 000.npy
│   │   ├── 001.npy
│   │   ├── 002.npy
│   │   ├── 003.npy
│   │   ├── 004.npy
│   │   ├── 005.npy
│   │   ├── 006.npy
│   │   ├── 007.npy
│   │   ├── 008.npy
│   │   └── 009.npy
│   └── labels
│       ├── 000.npy
│       ├── 001.npy
│       ├── 002.npy
│       ├── 003.npy
│       ├── 004.npy
│       ├── 005.npy
│       ├── 006.npy
│       ├── 007.npy
│       ├── 008.npy
│       └── 009.npy
└── valid
    ├── img
    │   ├── 000.npy
    │   ├── 001.npy
    │   ├── 002.npy
    │   ├── 003.npy
    │   ├── 004.npy
    │   ├── 005.npy
    │   ├── 006.npy
    │   ├── 007.npy
    │   ├── 008.npy
    │   └── 009.npy
    └── labels
        ├── 000.npy
        ├── 001.npy
        ├── 002.npy
        ├── 003.npy
        ├── 004.npy
        ├── 005.npy
        ├── 006.npy
        ├── 007.npy
        ├── 008.npy
        └── 009.npy
```
Raster Vision also expects this directory to be inside a ZIP file, so I then zipped it using (the `cd` trick is to prevent `data/` being the root dir inside the ZIP file) :
```sh
(cd data/ && zip -r ../data.zip ./)
```

Now to use this with Raster Vision, we will need a raster Vision config file. So create a new python file with the following contents.

Let's call it `config.py`:
```python
from rastervision.core.data import (ClassConfig, DatasetConfig)
from rastervision.core.rv_pipeline import (SemanticSegmentationConfig)
from rastervision.pytorch_backend import PyTorchSemanticSegmentationConfig
from rastervision.pytorch_learner import (SemanticSegmentationImageDataConfig,
                                          SemanticSegmentationModelConfig,
                                          SolverConfig, PlotOptions, Backbone)


def get_config(runner, out_dir: str) -> SemanticSegmentationConfig:
    class_cfg = ClassConfig(
        names=['background', 'target'],
        colors=['lightgray', 'red'],
        null_class='background')

    dataset_cfg = DatasetConfig(
        class_config=class_cfg, train_scenes=[], validation_scenes=[])

    data_cfg = SemanticSegmentationImageDataConfig(
        uri='./data.zip',
        img_channels=13,
        img_sz=256,
        num_workers=0,
        plot_options=PlotOptions(channel_display_groups={'RGB': [3, 2, 1]}))

    model_cfg = SemanticSegmentationModelConfig(backbone=Backbone.resnet50)
    solver_cfg = SolverConfig(lr=1e-3, num_epochs=3, batch_sz=4)
    backend_cfg = PyTorchSemanticSegmentationConfig(
        data=data_cfg,
        model=model_cfg,
        solver=solver_cfg,
        log_tensorboard=False)

    pipeline_cfg = SemanticSegmentationConfig(
        root_uri=out_dir, dataset=dataset_cfg, backend=backend_cfg)

    return pipeline_cfg
```

Now we can run this via:
```sh
rastervision run inprocess config.py train -a out_dir out/
```

You can now modify `config.py` to adopt it to your dataset. I suggest looking at the documentation for the following:
- [SemanticSegmentationImageDataConfig](https://docs.rastervision.io/en/latest/api.html#semanticsegmentationimagedataconfig)
- [SemanticSegmentationModelConfig](https://docs.rastervision.io/en/latest/api.html#semanticsegmentationmodelconfig)
- [SolverConfig](https://docs.rastervision.io/en/latest/api.html#solverconfig)

**Comment by nayemabs** [comment]
Thank you so much. I'll try this.

**Comment by AdeelH** [comment]
Closing this for now. Please feel free to reopen if you have further questions.
